android-components/components/concept/engine/src/test/java/mozilla/components/concept/engine/utils/EngineVersionTest.kt

```diff
@@ -68,6 +68,78 @@ class EngineVersionTest {
         assertEquals("68.3.0esr", "68.3esr".toVersion().toString())
         assertEquals("68.0.0", "68.0".toVersion().toString())
     }
+
+    @Test
+    fun `GIVEN a nightly build of the engine WHEN parsing the version THEN add the correct release channel`() {
+        val result = EngineVersion.parse("0.0.1", "nightly")?.releaseChannel
+
+        assertEquals(EngineReleaseChannel.NIGHTLY, result)
+    }
+
+    @Test
+    fun `GIVEN a beta build of the engine WHEN parsing the version THEN add the correct release channel`() {
+        val result = EngineVersion.parse("0.0.1", "beta")?.releaseChannel
+
+        assertEquals(EngineReleaseChannel.BETA, result)
+    }
+
+    @Test
+    fun `GIVEN a release build of the engine WHEN parsing the version THEN add the correct release channel`() {
+        val result = EngineVersion.parse("0.0.1", "release")?.releaseChannel
+
+        assertEquals(EngineReleaseChannel.RELEASE, result)
+    }
+
+    @Test
+    fun `GIVEN a different build of the engine WHEN parsing the version THEN add the correct release channel`() {
+        val result = EngineVersion.parse("0.0.1", "aurora")?.releaseChannel
+
+        assertEquals(EngineReleaseChannel.UNKNOWN, result)
+    }
+
+    @Test
+    fun `GIVEN an unknown release type WHEN comparing to other versions THEN return a negative value`() {
+        val version = "0.0.1"
+        val unknown = EngineVersion.parse(version, "canary")
+
+        assertEquals(0, unknown!!.compareTo(unknown))
+        assertTrue(unknown < EngineVersion.parse(version, "nightly")!!)
+        assertTrue(unknown < EngineVersion.parse(version, "beta")!!)
+        assertTrue(unknown < EngineVersion.parse(version, "release")!!)
+    }
+
+    @Test
+    fun `GIVEN an nightly release type WHEN comparing to other versions THEN return the expected result`() {
+        val version = "0.0.1"
+        val nightly = EngineVersion.parse(version, "nightly")
+
+        assertEquals(0, nightly!!.compareTo(nightly))
+        assertTrue(nightly > EngineVersion.parse(version, "unknown")!!)
+        assertTrue(nightly < EngineVersion.parse(version, "beta")!!)
```

Comment:
Wouldn't it be the opposite way? Nightly is newer so should be greater than beta?

Reply:
I think this ends up just comparing the values in the ReleaseChannel enum because of the `compareTo` changes above. In reality, Nightly will have a greater major version anyway, yielding a different result.

A Nightly of the same major version would be older than beta, so this seems correct? A Nightly of a higher major version would be newer. Maybe that's a test we should be adding :)

Reply:
Right! I should have make it more clear that the reason it was giving us this result is because of the (unintended?) side-effect of the enum positioning.

Reply:
The enum comparison was intended with the members being arranged in a preference order with the last being preferable to the ones above.
I went with thinking about two jars, with version `100.0.0`, one from Nightly, one from Release.
The `100.0.0` - `release` jar is newer (by ~2 months) and also (hopefully) more stable so for the same exact version the `release` jar is preferable and a comparison will return a positive result.

Thank you for the suggestions about a new test! In the current code the release channel comparison is done last so I didn't think a test is needed for different versions and different release channels since we already have tests for different versions but that code can be refactored and so a test would help catch regressions.